src/pam/mod.rs

```diff
@@ -301,19 +330,22 @@ impl<C: Converser> PamContext<C> {
                 res.insert(k, v);
             }
 
-            // free the current string and move to the next one
+            // SAFETY: curr_str was obtained via libc::malloc() so we are responsible for freeing it.
```

Comment:
One additional thing we might note here: we never let the pointers from curr_env escape the above loop, instead creating an owned copy by running `to_owned()`, so there should be nothing pointing to anything related to curr_env/curr_str.

Reply:
It would be cool to have Rust check this for us by grabbing a `&mut` instead of a `NonNull`. Sadly a `NonNull` is `Copy`. Perhaps we can change it earlier on in something that isn't.

Reply:
Addressed in e569638.

Reply:
Note: this PR is now rebased on top of #867, which should be merged first as it is easier to review. :)